### PR TITLE
Ampers: Kirsten Schumy

### PR DESCRIPTION
**Pull request by kschumy**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I started with a lot of the same validation method in different classes. I realized I could change these into module methods. This helped a lot with testing because I could test those method and not have to worry about testing them individually in the classes that used them. 
Describe any examples of composition that you encountered in this project, if any | Trip has a passenger and has a driver
Describe a _nominal test_ that you wrote for this assignment. | I tested that driver could add a normal, not in-process trip and that it would calculate cost and time
Describe an _edge case test_ that you wrote for this assignment. | I made sure that trips with costs below the ride fee did not result in the driver owing money
Describe a concept that you gained more clarity on as you worked on this assignment. | The module methods idea and working with objects and references

### CODE REVIEW

**Comment by kschumy** [comment]
I submitted it to the wrong one. Sorry! I'll delete the other one

File: lib/driver.rb [hunk 1]
```diff
@@ -1,47 +1,147 @@
...
+    end
 
-      if trips.length == 0
-        average = 0
-      else
-        average = (total_ratings.to_f) / trips.length
+    # Provided trip_cost must be the cost of a trip in dollars.
+    # Returns the amount in dollars the
+    def calculate_pay(trip_cost)
+      return trip_cost >= TRIP_FEE ? (trip_cost - TRIP_FEE) * 0.80 : 0.0
```
Comment (CheezItMan) [inline lib/driver.rb:107]: Good use of the constant and ternary operator.

File: lib/passenger.rb [hunk 2]
```diff
@@ -3,22 +3,62 @@ class Passenger
...
+    end
+
+    # Returns the
+    def get_total_money_spent
+      return calculate_total_money_spent
+    end
+
+    # Returns the sum of all complete trips durations in seconds.
+    def get_total_time
+      return RideShare.get_all_trip_durations_in_seconds(@trips)
```
Comment (CheezItMan) [inline lib/passenger.rb:30]: This works, it might be better as a mix-in, which we haven't studied.

File: lib/rideshare_module_methods.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
+# Stores the module methods
+module RideShare
```
Comment (CheezItMan) [inline lib/rideshare_module_methods.rb:2]: Nice idea, I think these make more sense as a mix-in but you definitely have the right idea.

File: lib/trip.rb [hunk 4]
```diff
@@ -1,21 +1,100 @@
...
+    end
+
+    # If trip is not in-progress, returns the trip duration in seconds.
+    # Otherwise returns 'nil'.
+    def get_duration
+      return (@end_time - @start_time).to_i if !is_in_progress?
+    end
+
+    # Returns 'true' if the trip is in progress and 'false' otherwise.
+    def is_in_progress?
```
Comment (CheezItMan) [inline lib/trip.rb:27]: 👍

File: lib/trip_dispatcher.rb [hunk 5]
```diff
@@ -15,87 +15,155 @@ def initialize
...
+      trip = Trip.new(new_trip_data)
+      trip.driver.add_trip(trip)
+      trip.passenger.add_trip(trip)
+      return trip
+    end
+
+    # Returns an available driver, with priority going to the driver who has
+    # never had trips and then to the driver with the least recent last trip.
+    # Returns nil if cannot find an available driver.
+    def find_available_driver
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:144]: Nicely done!

**Comment by CheezItMan** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Good number of commits and good commit messages
Answer comprehension questions | Check, good edge-case test example
**Wave 1** |
`Trip` has a helper method to calculate duration | Check
`Passenger` has a method to calculate total cost of all trips | Check, interesting use of a method in the RideShare module.  Good reuse of code.
`Driver` has a method to calculate total revenue | Check
Complete tests |  Check
**Wave 2** |
`TripDispatcher` has a new method to create trips |  Check
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | Check
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | Check
Complete tests | Well done
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | Check
Appropriate use of Ruby's `Time` |  Check
Appropriate helper methods were made to help with complex logic | Well done breaking the method down.
Complete tests | Check
**Overall** | Outstanding work.  You hit all the learning goals including wave 3.  I like how you created reusable methods outside the classes.  That shows you're coming at things from the right angle.
